proposals/type-classes.md

```diff
@@ -0,0 +1,298 @@
+```
+
+The below alternative approach to `given` using parameters and the special keyword `instance` was also proposed but discarded since `given`
+was more inline with other similar usages such as `where` that users are already used to and did not require to name the instances to activate extension syntax.
+
+```kotlin
+class OptionMonoid<A>(instance MA: Monoid<A>) : Monoid<Option<A>> //class position
+
+fun <A> add(a: A, b: A, instance MA: Monoid<A>): A = a.combine(b) //function position
+```
+
+## Compiler Changes
+
+- The type checker will declare the below definition as valid since the `given` clause provides evidence that call sites won't be able to compile calls to this function unless a `Monoid<A>` is in scope.
+```kotlin
+fun <A> add(a: A, b: A): A given Monoid<A> = a.combine(b) //compiles
+```
+- The type checker will declare the below definition as invalid since there is no `Monoid<Int>` in scope.
+```kotlin
+add(1, 2)
+```
+- The type checker will declare the below definition as valid since there is a `Monoid<Int>` in scope.
+```kotlin
+import intext.IntMonoid
+add(1, 2)
+```
+- The type checker will declare the below definition as valid since there is a `Monoid<Int>` in scope.
+```kotlin
+fun addInts(a: Int, b: Int): Int given Monoid<Int> = add(a, b)
+```
+- The type checker will declare the below definition as valid since there is a `with` block around the concrete `IntMonoid` in scope.
+```kotlin
+fun addInts(a: Int, b: Int): Int = with(IntMonoid) { add(a, b) }
+```
+
+## Compile resolution rules
+
+When the compiler finds a call site invoking a function that has type class instances constrains declared with `given` as in the example below:
+
+Declaration: 
+```kotlin
+fun <A> add(a: A, b: A): A given Monoid<A> = a.combine(b) 
+```
+Call site:
+```kotlin
+class AddingInts {
+  fun addInts(): Int = add(1, 2)
+}
+```
+The compiler may choose the following order for resolving the evidence that a `Monoid<Int>` exists in scope.
+
+1. Look in the most immediate scope for declarations of `given Monoid<Int>` in this case the function `addInts`
+
+This will compile because the responsibility of providing `Monoid<Int>` is passed unto the callers of `addInts()`:
+```kotlin
+class AddingInts {
+  fun addInts(): Int given Monoid<Int> = add(1, 2)
+}
+```
+
+2. Look in the most outher class/interface scope for declarations of `given Monoid<Int>` in this case the class `AddingInts`:
+```kotlin
+class AddingInts given Monoid<Int> {
+  fun addInts(): Int = add(1, 2)
+}
+```
+This will compile because the responsibility of providing `Monoid<Int>` is passed unto the callers of `AddingInts()`
+
+3. Look in the import declarations for an explicitly imported instance that satisfies the constrain `Monoid<Int>`:
+```kotlin
+import intext.IntMonoid
+class AddingInts {
+  fun addInts(): Int = add(1, 2)
+}
+```
+This will compile because the responsibility of providing `Monoid<Int>` is satisfied by `import intext.IntMonoid`
+
```

Comment:
I would like `*` imports. The only issue with `*` imports is that if the compiler found ambiguous instances it should bail with a proper message. It was suggested at some point that those so be explicit so that newbies were not confused as to where the instances where getting applied from but if we want to support `import kategory.*` for example to bring all of our instance into scope that would be awesome. The IDE is doing something similar in the Scala plugin in the case of implicits. It looks in all symbols imported in a given scope trying to find candidates for resolution and it if it finds one it activates the syntax.

Reply:
`*`-imports will likely pose a performance challenge for the compiler

Reply:
Something worth mentioning is that instances are only resolved at call sites where the invocation is concrete so the compiler does not need to look into all `*` where they are declared just where functions are invoked. I believe is the same way resolution now works to bring other symbols into scope without fully qualifying them with the full package name.

Reply:
> just where functions are invoked.

It's very-very many places in the code :)

> I believe is the same way resolution now works to bring other symbols into scope without fully qualifying them with the full package name.

Not quite. Other symbols are bound by name, here we are binding by type, and it's a lot more work for the compiler

Reply:
What about cases where you want to import two implementations of the `Monoid<Int>` into the scope of one file?

Do you have to resolve this conflict by calling each scoped by a `with` block??

```kotlin
import intext.IntMonoid1
import intext.IntMonoid2

fun addInts1(a: Int, b: Int): Int = with(IntMonoid1) { add(a, b) }
fun addInts2(a: Int, b: Int): Int = with(IntMonoid2) { add(a, b) }
```